frame/node-permission/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Node permission moudule
+//!
+//! This module is used by the `client/service` to retrieve the current
+//! permissioned nodes.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use sp_std::fmt::Debug;
+use sp_core::storage::well_known_keys;
+use sp_runtime::traits::{Member, MaybeSerializeDeserialize, MaybeDisplay};
+use frame_support::{
+    decl_module, decl_storage, decl_event, decl_error,
+    Parameter, storage, ensure,
+    weights::{DispatchClass, Weight},
+    traits::{Get, EnsureOrigin},
+};
+use codec::Encode;
+
+pub trait WeightInfo {
+    fn add_node() -> Weight;
+    fn remove_node() -> Weight;
+    fn swap_node() -> Weight;
+    fn reset_nodes() -> Weight;
+}
+
+impl WeightInfo for () {
+    fn add_node() -> Weight { 50_000_000 }
+    fn remove_node() -> Weight { 50_000_000 }
+    fn swap_node() -> Weight { 50_000_000 }
+    fn reset_nodes() -> Weight { 50_000_000 }
+}
+
+pub trait Trait: frame_system::Trait {
+    /// The event type of this module.
+	type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+
+    /// The maximum number of permissioned nodes that are allowed to set
+    type MaxPermissionedNodes: Get<u32>;
+
+    /// The node identifier type for the runtime.
+    type NodeId: Parameter + Member + MaybeSerializeDeserialize + Debug + MaybeDisplay
```

Comment:
Data that is stored under a well known key should not be generic. The data should be well known to be decodable by the client side.

Reply:
Libp2p has a few options for a peer's public key, below I refer to its docs:
```
/// The public key of a node's identity keypair.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum PublicKey {
    /// A public Ed25519 key.
    Ed25519(ed25519::PublicKey),
    #[cfg(not(target_arch = "wasm32"))]
    /// A public RSA key.
    Rsa(rsa::PublicKey),
    #[cfg(feature = "secp256k1")]
    /// A public Secp256k1 key.
    Secp256k1(secp256k1::PublicKey)
}
```
Substrate network now only use ed25519 for our connection, we may need to support to other type (would be good to get confirmation from @tomaka , sorry to ping you again 😄 ). Does it makes sense to make it configurable in this scenario? The runtime composer will decide which node key to use, kind of more friendly to the network upgrade of other type supporting.

Reply:
I would put something like this in the runtime code:

```rust
#[derive(Encode, Decode)]
pub enum NodePublicKey {
    Ed25519([u8; 32]),
}
```

This makes it extensible for the future.

Then convert from that `NodePublicKey` to a `libp2p::PublicKey` in the service.

I'd name this `NodePublicKey` and not `NodeId` to make it clear that it's a public key and not a `PeerId`.

Reply:
thanks @tomaka  here is my two cents:
* I'm not a fan of adding more `Ed25519` definition since we already have similar type in sp-core and libp2p. 
* Once network is launched, there should be only 1 type of key to make the connection. Adding such enum seems over engineering, and make user input to add a permissioned node less friendly.

Let me know what do you think.

Reply:
> I'm not a fan of adding more Ed25519 definition since we already have similar type in sp-core and libp2p.

I'm the opposite opinion: types should be scoped for specific usages. The ed25519 definition in libp2p corresponds to what libp2p is concerned about. The ed25519 definition in the runtime corresponds to what the runtime is concerned about. Just because they share a name doesn't mean that they should be merged into one type.

> and make user input to add a permissioned node less friendly.

I don't think we should make the code less future-proof just because it would make it more friendly, if that's what you suggest.
This `NodePublicKey` represents how the runtime communicates with the client. It doesn't mean that users should directly manipulate this enum. We can put whatever user-friendly API we want a top.

Reply:
Now I get your points. 
I'm imaging to make a configurable NodePublicKey in this pallet, and define this enum in sp-runtime (or sp-core, I need to check the difference, similar to how AccountId is defined), is this the same structure in your thought?
Also I think this refactor can be introduced in another PR, so that people can review it more clearly, does it makes sense?

Reply:
> a configurable type

What do you mean by "configurable"?
All we really need is a type that we can potentially extend in the future with new types of keys.

> Also I think this refactor can be introduced in another PR

I'm extremely way of this. Not that I don't trust you, but in practice this type "I'll do another PR later" tend to never happen.
Is it really so big of a change that it can't be done in this PR here?

Reply:
> What do you mean by "configurable"?

Similar to how `AccountId` is defined in frame-system, [code](https://github.com/paritytech/substrate/blob/master/frame/system/src/lib.rs#L208).

> Is it really so big of a change that it can't be done in this PR here?

I want to put the `NodePublicKey` enum in sp-runtime and not sure how others feel about it, it may involve more discussion. This PR already got a lot of discussion, so I think split the further work to make a bit clean. Here I may follow a practice that not fit for this repo though.